### ISSUE TITLE
Allow users to specify an IP for their container

### ISSUE THREAD

**Issue by vishh** [docker/docker#6743]
There is no way to tie a container to an IP in docker as of now. #2801 captures this limitation.
If users were allowed to specify an IP in 'docker0' subnet that would solve this issue.
With #6704 users can be specify IPs even outside the default docker0 bridge.
https://github.com/dotcloud/docker/pull/5829 was the first attempt to solve that issue. 

+@crosbymichael

**Comment by acornejo** [comment]
+1

**Comment by crosbymichael** [comment]
@erikh 

just so this is not forgotten

**Comment by erikh** [comment]
tyvm

**Comment by jessecooper** [comment]
+1

**Comment by pgenevski** [comment]
+1

**Comment by coolljt0725** [comment]
+1

**Comment by Kokonoe** [comment]
+1

**Comment by mgcrea** [comment]
:+1:

**Comment by Ovski4** [comment]
+1

**Comment by AndreiMuresan** [comment]
+1

**Comment by hansgru** [comment]
+1

**Comment by mgcrea** [comment]
My alternative to address this has been to use `docker-gen` to regenerate on the fly the /etc/hosts of my host, providing avahi-like hostnames `my-container-name.docker`.

**Comment by nfnty** [comment]
+1

**Comment by re6exp** [comment]
+1

**Comment by xiaods** [comment]
anyone can clarify the detail proposal. do we need implement it in docker run args? i am also interest this feature implement.

**Comment by chenchun** [comment]
+1

**Comment by dchen1107** [comment]
+1

**Comment by yinheli** [comment]
+1

**Comment by hilyjiang** [comment]
+1

**Comment by barjomet** [comment]
+1

**Comment by xiaods** [comment]
today review the original comments and @shykes 's comments, currently we need know where is the plugin codebase, then we can evaluate the ip feature in this new plugin architect.  ping @erikh , do you know where is best start plugin code base?

**Comment by feiskyer** [comment]
+1

**Comment by cfrendo** [comment]
+1

**Comment by xiaods** [comment]
@thaJeztah anyone can help to maintain this feature, give some feedback from docker maitainers?

**Comment by thaJeztah** [comment]
@xiaods 

I'm not "in the loop" myself, but have seen some activity in this area (no actual *code* yet), 

I think there's currently being worked on a proposal for the "plugin API" in Docker.

For what I've seen, several members of Docker *and* projects in the Docker "ecosphere" are discussing the possibilities to reach (initial) consensus on an API that works, both for Docker itself (default plugins) and third-party projects.

The preparations/discussions for this proposal are (not yet) taking place here in this repository; you should be able to find it, but I think (just assuming this) the parties involved want to be able to have a "quiet" discussion first, before the *whole* community is involved. :smile:

The plugin API would "pave the road" for more features, for example, plugins for "volumes" and "networking" (such as proposed in https://github.com/docker/docker/issues/9983)

Also, this [press release](http://www.businesswire.com/news/home/20150304005595/en/Docker-Acquires-SocketPlane-Drive-Open-Collaborative-Networking#.VSeNlUs9InV) and [blog-post](http://blog.docker.com/2015/03/socketplane-excited-to-be-joining-docker-to-collaborate-with-networking-ecosystem/), gives indication that there's more to come :)

*Note that the above is just information I collected by watching this repository.* I'll ask around if there's a maintainer that is able to provide more information and, if possible, a status report (or estimation).

**Comment by jhalterman** [comment]
+1

**Comment by jaarandaz** [comment]
+1

**Comment by SerCeMan** [comment]
+1

**Comment by nicktgr15** [comment]
+1

**Comment by astlock** [comment]
Can't switch to libcontainer from LXC exec-driver because of this.

**Comment by thaJeztah** [comment]
If all goes according to plan, Docker networking will be replaced with ["libnetwork"](https://github.com/docker/libnetwork) in Docker 1.7 (also see [the new blog post](http://blog.docker.com/2015/04/docker-networking-takes-a-step-in-the-right-direction-2/)

Initially (afaik), this will be only a "behind the screens" change, without new features added, but new features may come in future releases.

To check progress on libnetwork (and they'll be looking for contributors/help soon), see the libnetwork repository; https://github.com/docker/libnetwork

@mrjana will this be supported once libnetwork is integrated?

**Comment by mrjana** [comment]
@thaJeztah Yes this is something we are considering to support in libnetwork(after the initial integration ofcourse) using well-defined labels that the drivers are required to honor or reject mandatorily. Much the same way as `--mac-address` configuration is handled right now in libnetwork.

**Comment by thaJeztah** [comment]
@mrjana sounds good! Is there an issue in place already in libnetwork that people can track (what's the best place for that, in libnetwork or here?)

**Comment by mrjana** [comment]
@thaJeztah There is no issue tracking it, so yes it should be filed in libnetwork

**Comment by xiaods** [comment]
i feel the request can be close right now. all more request can routing to libnetwork

**Comment by thaJeztah** [comment]
@xiaods I think it should be left open for now (people will search for open issues in the Docker issue tracker, not in the Libnetwork repository)

however, if you could open an issue in the Libnetwork issue tracker and add a link here, that would be great!

**Comment by mjaverto** [comment]
+1

**Comment by basickarl** [comment]
+1

**Comment by chiavegatto** [comment]
:+1:

**Comment by chiavegatto** [comment]
+1

**Comment by itoffshore** [comment]
+1

**Comment by thaJeztah** [comment]
in stead of adding +1s, could someone open an issue in the https://github.com/docker/libnetwork repository (if it doesn't exist yet). For reference, see https://github.com/docker/docker/issues/6743#issuecomment-99090213

**Comment by mouson** [comment]
+1

**Comment by Alives** [comment]
Look at all these people who don't understand docker or github... so tired of seeing these updates to an issue I chose to follow... maybe I'll just rely on my linking and stop caring.

-1  learn docker, learn github, stay out of my inbox.

**Comment by MarkMielke** [comment]
The +1 aren't valuable... definitely it's annoying after a time and I'd like people to stop as well. But I don't follow the rest of your comment. I'm not sure whether to parse the -1 as being against this feature request, or against the people who would -1.

Presuming it is for the feature request, I felt the need to state that it is from learning more about Docker that I reached the conclusion that dynamic assignment of IP addresses meets a certain set of requirements, but it fails a different set of requirements. The dynamic assignment of IP addresses has lead to us using host networking more than we would like. This is for various reasons including avoiding the overhead of NAT, and preventing us from loosely linking services together. The "container linking" model (with some improvements) would work well to an extent, but not every problem fits into the domain that "container linking" is designed to address.

I use both. NAT of dynamic private addresses for services that are well suited to this, and host networking for services that are not well suited to this. If we had the ability to set up static IP addresses for containers, this would let us define a host network interface as a bridge, and allow the Docker containers to be accessed directly without the need for NAT.

One feature we have mostly lost moving applications over to Docker is the ability to effectively use iptables of our own along-side the docker maintained iptables. Docker does try to be friendly by limiting its use of iptables, but most packages that work with iptables are not Docker friendly. So this is another place where host networking vs NAT with private addresses becomes a problem for us.

Now, I'm sure if static IP is made available that a number of people will use this feature inappropriately. Perhaps they should "learn Docker" and try to do it the right way, which might not be static IP, but Docker should provide the full feature set to meet all of the requirements and let us decide and make the right or wrong decision. It isn't for somebody in an ivory tower considering only their own use cases and their own understanding of what requirements might be, to dictate how everybody else should be limited.

That's my late night rant for today... not sure how much is applicable or not. :-)

**Comment by eolamey** [comment]
@MarkMielke I am not saying that would solve the need for static IP addresses, but you can already disable (source and destination port) NAT for the docker daemon, add one of your host interface to the docker bridge, and have docker assign an IP address to containers out of a pool that is routable on your network. You don't even have to assign a whole subnet to each docker host, a subset would do.
Static addresses and "private" (non routable) addresses are two different issues.

**Comment by itoffshore** [comment]
Temporary solution for [assigning static ip addresses to Docker containers](https://it-offshore.co.uk/linux/debian/36-giving-docker-containers-static-ip-addresses).

**Comment by MarkMielke** [comment]
@eolamey You are correct. I might not have been clear that I was talking about the need for static IP address, where the current options are:

1) Use NAT + port mapping, from a static IP address.
2) Use host networking.
3) Use a temporary solution. Pipework was the one I was originally look at. The temporary solutions don't perfectly fit. For Pipework in particular, I saw a race condition during setup and the general recommendation not to use it by its own author as good reasons to avoid it. I understand how it works (I think), and it's a really great option if you really need this functionality while it doesn't exist in Docker itself, but I also have the option of falling back to host networking, and when it comes to recommending which solution is the safest, the most robust, and the easiest to support until Docker does come with this functionality out of box, I am currently electing to go with host networking. This still gives us substantial value over the previous model, and for the use cases that is applicable, we don't (yet) need scale out with multiple instances running in parallel and other such features.

So, being able to have Docker dynamically assign from a routable address space doesn't really solve any problem I have right now that I am aware of. It needs the static IP address capability in concert to really be useful for my use cases.

**Comment by thaJeztah** [comment]
There's now a libnetwork issue/ feature request that can be found here: https://github.com/docker/libnetwork/issues/161

**Comment by JonasT** [comment]
This is crucial for IPv6, or how else am I supposed to set the DNS of my web domains to my actual docker machines when my docker containers' ips just randomly change?? Since you didn't bother to include a NAT for IPv6 either, it would be nice if this could finally be addressed in a timely manner.

**Comment by dikerboyz** [comment]
Hi, could you tell me how to use pipework to assign docker container IP Address?

**Comment by thaJeztah** [comment]
@dikerboyz the issue tracker is not really the place for discussing that. Look at the [pipework documentation](https://github.com/jpetazzo/pipework), or ask your question in the #docker IRC channel, or [Docker Users](https://groups.google.com/forum/#!forum/docker-user)

**Comment by itoffshore** [comment]
@dikerboyz - see this [temporary solution for static ip addresses in docker](http://it-offshore.co.uk/linux/debian/36-giving-docker-containers-static-ip-addresses). I've also written a [script](https://gist.github.com/itoffshore/9e30f6c857ab51f05d9b) to make it easier.

**Comment by JonasT** [comment]
@itoffshore and sadly it's quite complicated (compared to just adding a parameter to docker run/docker-compose.yml), which is why this really needs to be added to mainline docker.

**Comment by itoffshore** [comment]
@JonasT - I've not needed to use `ipv6` - but [pipework does support ipv6](http://stackoverflow.com/questions/28495222/docker-1-5-on-ipv6-only-host)

**Comment by JonasT** [comment]
Why does nobody need ipv6? ipv4 addresses are running out: http://arstechnica.com/business/2015/05/get-ready-to-wait-in-line-for-more-ipv4-addresses/

**Comment by pdumais** [comment]
+1

I would need this. I've hacked my way through this by adding a "--ipv4-address" param
https://github.com/pdumais/docker

Dirty hack but it works perfectly for me. Can't wait to get this officially.

**Comment by yinheli** [comment]
@pdumais Greet Job +1

**Comment by iano123** [comment]
+1

**Comment by brunowego** [comment]
+1

**Comment by candlerb** [comment]
I note that func `RequestIP` in `vendor/src/github.com/docker/libnetwork/ipallocator/allocator.go` already takes an optional argument which is the address you want to be used.

So it seems to me all that's needed is a command-line argument to `docker run/start` which is passed through?

(This is assuming that people are happy to specify the IP each time the container starts, and not have it persist through stop/start cycles. The latter would required the allocator to take note of all addresses assigned to stopped containers)

**Comment by pdumais** [comment]
@candlerb That's basically what I did there: https://github.com/pdumais/docker
It works fine for my needs.

**Comment by danlamanna** [comment]
+1

**Comment by psi-4ward** [comment]
@candlerb very nice and could be so easy

**Comment by duobei** [comment]
+1

**Comment by hasufell** [comment]
++

**Comment by cHolzberger** [comment]
+1

**Comment by qm78** [comment]
+1

**Comment by nhutdm** [comment]
+1

**Comment by mavenugo** [comment]
Thanks for all the +1s that keeps the interest alive.
am trying to parse through all the discussions across various PRs to get to a solution that satisfies the asks. I see a few different requests in these.

1. ip-address stability for the containers across container restarts and daemon restarts. In this requirement, the users really doens't care what the specific IP for the container is, but would like to keep it stable across restarts. With the new CNM model, this is a simple fix & can be turned on anytime. But, retaining the IP of a container when it stops will lead to ip-address exhaustion if it is not used responsibly. Yes, it is an user problem much like disk-space. So my personal vote is to provide the ip-address stability by default.

2. Ability to assign predefined and static-ip-address to the container. This particular request brings in the operational design questions as raised here : https://github.com/docker/docker/pull/13476#issuecomment-109363464 . So, the real question is, do we really care about assigning a specific IP address to a container or achieve the same via a more generic operations friendly interface such as IPAM driver which can parse specific labels to assign ip-address (as explained by @shykes above ? ). implementing a `--ip` in docker client is simple, but, it would be great if we can avoid adding such knobs for individual options. Rather try to find a solution that works for both developer and operations need. We are currently exploring the ability to provide pluggable IPAM driver which users can make use of to provide "preferred" ip-address  to their network/containers/services.

3. NAT and not. With the new CNM model (contributed from docker/libnetwork), the users can choose from a variety of network plugins that satisfies the need. Even better, user could write an external networking plumbing driver and plug into the system that will satisfy their specific requirement. This feature is currently in the experimental channel https://github.com/docker/docker/blob/master/experimental/networking.md, https://github.com/docker/docker/blob/master/experimental/plugins_network.md.

It will be very useful if everyone who +1 this proposal can give your input, so that we can prioritize the work appropriately.

**Comment by qm78** [comment]
@mavenugo I appreciate your summary

My use case for a fixed container IP is, that I've dockerized a samba4 AD DC as common authentication backend for my other containers. Now I'd like to set its IP as dns server on my other containers run (--dns parameter). 
My current workaround is to expose samba ports to my host and set host's IP as dns server for my other containers. This approach works, but I think it's a workaround, because it breaks isolation and opens port to host, which I normally wouldn't have opened. 
I'm sure container linking also would work, but then I'd manually need to modify my containers' resolv.conf on run.

So obvious solution would be to run a container with fixed IP, even it violates docker's concepts and best practices.

Probably most sophisticated way would be an option to tell a container on run the DNS providing container based on its name. eg.: docker run --dns-from-container {container name} ...

As a docker newbie still discovering docker universe I apologize in case I oversaw something and I'm looking forward to your input.

**Comment by cpuguy83** [comment]
@qm78 combo of `--link` + `--dns` seems to be workable here.
`docker run --link smb4:dns --dns dns`

**Comment by candlerb** [comment]
I am in group (2), and yes, I really care about assigning a specific IP address.

Use case: I want to set up a bunch of docker containers as a virtual lab, and in order that the exercises match the documentation, the containers need to come up on pre-defined IP addresses.  Optionally the address to use can be persisted with the container metadata (like case (1)), but the actual address to use needs to be selectable - i.e. if the IPAM is used then I need a flag to tell the IPAM exactly which IP address to use.

This is essentially the "use as container as a lightweight virtual machine" case.

For networking I would be attaching to a bridge without NAT, i.e. `DOCKER_OPTS="-b br-lan --ip-masq=false --iptables=false"`

If the container name resolved to an IP address via DNS, that would also work. I would be naming my containers using their DNS names anyway for consistency.

Note that as it stands, this is still restricted in that all containers have a single NIC, and all NICs must attach to the same bridge.  There are some VM cases where I'd like this to be more flexible. Specifying multiple NICs in JSON would be straightforward, but as command line flags it could be ugly, e.g..

~~~
docker run --net iface=eth0,bridge=br-lan,v4=10.0.0.1/24,gateway=10.0.0.254 \
           --net iface=eth1,bridge=br-wan,v4=10.1.0.1/24 foo
~~~

For a particular NIC you might want to auto-assign an address via IPAM (`v4=auto` ?) or leave it unset for the container to do its own dhcp (`v4=none` ?)

At this point I can see why people might say that it's better to have something else configure the NIC (like pipework).

So an alternative solution would be if `docker run` would invoke a pluggable network configuration script, and pass through a blob of configuration data (which ideally persists with the container). This is for two reasons: (a) to make `docker run` more convenient, and (b) so that the networking inside the container is setup *before* the container starts running.

**Comment by psi-4ward** [comment]
My use case for static IPs is to replace the old VMs in a multi-host environment.
Currently i add an extra physical interface to docker0 (ovs bridge) to bringe up the private lan between my hosts.

Static IPs are really essential for me:
* No IP conflicts caused using dockers host-specific *"dhcp"*
* No (unstable) discovery magic (IPAM, Skypdns, Registrator, whatever)
* IPTable rules and/or VLANs to encapsulate container groups 
* More robust applications setups, not every App re-resolves Hostnames. Some Apps don't event support hostnames, you need to specify the IP. This leads to an App restart using dynamic IPs when any of its services gets restarted. Have fun with implementing a controller doing this.
* Easy config for infrastructure services like Monitoring or Loadbalancers

My current solutions is starting a container with `--net=none` and use pipework to assign the IP and default gateway:
* I have to wait an unknown time until i can start pipework. Docker needs some ms to *start* the container.
* Some apps dont bind to an iface that was added after the app was started: Container patching needed with `pipework --wait`
* docker stats for net I/O doesn't work with pipework

Bottom line: We need a way to specify the IP,Mask and Route (at least for eth0) before the entrypoints is reached.

**Comment by qm78** [comment]
@cpuguy83: tried your approach, but received error telling  that --dns expects an IP address :-(
<code>invalid value "dc01" for flag --dns: dc01 is not an ip address</code>
Tested with: Docker version 1.8.1, build d12ea79

Any other idea, how to solve? Hmm either --dns flag could be relaxed to accept container names and to implicitly lookup target container's IP or to introduce a new flag --dns-from-container as proposed in my last post ... or allow static container IP. What do you think?

**Comment by patelrit** [comment]
We have the same use case as @qm78. Having static container IP or providing container name in the --dns flag would work.

**Comment by JonasT** [comment]
I just want to use IPv6. Therefore the container needs to either have a predictable, static IP (not the changing nonsense it has right now) or IPv6 needs to be behind a NAT too. I don't really care much which of those two it's gonna be.

**Comment by misiek08** [comment]
In my case specifying IP by argument like -ip looks like best solution. I have some machines with /24 networks and starting container makes few routing rules in iptables (each container need 4 ports - redis and 2 daemons - don't tell me to run redis and daemon in separate containers, my arch need current approach).

IPAM drivers and other stuff can be done and many will use them, but just setting static IP it's best way I can imagine for now in this issue. DHCP can be resolved other way and IPAM and other solutions will take much more time than simple static address.
"Bottom line: We need a way to specify the IP,Mask and Route (at least for eth0) before the entrypoints is reached. " +1

**Comment by drobison00** [comment]
+1
I've got a number of services that do not allow configuration using hostnames, and are not likely to be changed anytime soon. Its extremely frustrating that Docker still doesn't support something like an --ip flag, especially since this was available under the LXC model.

**Comment by netroby** [comment]
+ 65536

Want docker allocate ip address then remember the ip for one container.
Container restart will not change the default ip allocated.

If removed container. then release the ip .

for example, we create an MySQL DB container,

Other web front app, will connect the DB container for retreive data and storage data.

We will not want to change the DB service ip address.

we want it persist.

**Comment by thaJeztah** [comment]
@netroby if the "web front app" is also running in a container, then using `--link yourmysqlcontainer:db` or the new (still experimental) networking and Services (see [experimental/networking.md](https://github.com/docker/docker/blob/81cc8ebc93c20537d5fab4c016a3774f58b3e483/experimental/networking.md)) could be what you're looking for

**Comment by Frogging101** [comment]
:+1: 
This is really my only major issue with Docker. I run multiple services on different IP addresses, and there's far too much nonsense involved in routing a specific container to a specific public IP. I've taken to setting custom MAC addresses on my containers and fwmarking the packets with ebtables so that the host knows where to route them. It works, but it's pretty ridiculous that I have to kludge around like that :P

And you can just forget about using IPv6 because of the random address assignments.

**Comment by ekoyle** [comment]
I would also like to see a way to handle static IP address assignment to a host.

I was hoping to use docker to containerize our IPAM systems (DHCP, DNS, radius, ...).  These services are by their very nature IP address dependent (I have to manually configure my equipment to connect to a static IP).  I use different IP addresses for each service so that I can move services around, since I have no desire to run every service on every address.

Currently, to get this to work in docker, I would have to add the external IP address on my host and then let docker configure NAT rules in iptables to do an extra translation step which, although flexible, is adding a layer that a) I don't need, b) adds a lot of manual configuration, and c) is much more difficult to debug when something goes wrong.

If the container owned the external IP address, I could easily check to see if a container was running based on whether it responded to low-level protocols (ie. arp, icmp echo-request, etc.) remotely.  It also simplifies troubleshooting because I don't have to try to figure out what iptables might be doing to the packets once they hit the host.

Static IP addresses seem to be the cleanest solution here.

**Comment by thaJeztah** [comment]
@mavenugo is this something that can be "resolved" now that we have [pluggable IPAM driver support in 1.9](https://github.com/docker/docker/pull/16910)?

**Comment by mavenugo** [comment]
@thaJeztah based on an earlier polling I conducted, I got mixed requirement from many users.
1. some just want to assign ip-address to container via `run`, 
2. some wanted ip-address stability across restarts, 
3. others are operations focused who want to take complete control over the IPAM (such as DHCP and other IP Address Management solutions).

#16910 solves #3 above. This IPAM support is for `docker network` which enables the users to take control over the IP Assignment on that `docker network` (without having to deal with per-container static ip-address assignment).
It has pluggable IPAM driver support and hence lets the user to write their own drivers / or reuse other drivers in order to manage the network.

So, if the above solution satisifies the requirement defined in this PR, we can close it.

But, if the requirement is about letting user specify a single static ip-address for a container, we dont have an UX for that yet.

**Comment by JonasT** [comment]
I'm reiterating to make sure this isn't forgotten: it should be possible to specify *both* an IPv4 and an IPv6 for a container. Just one single generic "ip" field for one IP is not enough for a modern dualstack environment.

The IPAM solution is not very helpful because a main advantage of docker is how easy it is to get started. If I understand correctly it's quite a complex approach for entirely custom network configuration, which is way beyond what is required for most simple use cases that involve setting a specific IP.

**Comment by mavenugo** [comment]
@JonasT agreed. thats the reason I mentioned IPAM as operations focussed.
BTW, by default 1.9.0 will use the in-build IPAM solution which supports both ipv4 & ipv6.
User will be able to choose their subnet | ip-range of choice per network and the containers connect to the network will have ip/ipv6 allocated from that subnet | ip-range.

Pls refer to https://github.com/docker/docker/pull/7436#issuecomment-61350487 for more discussion around operations vs dev concerns.

**Comment by Frogging101** [comment]
I'm still waiting until there's a simple way to set static custom IP addresses on individual container instances.

**Comment by itoffshore** [comment]
For fixed ip addresses I've been using [Alpine Linux](http://alpinelinux.org/) LXC containers with `runit` to manage the services & logging by `socklog`. 

If you `apk add runit runit-doc socklog apk-post-messages` the Alpine `README` will be printed with instructions to configure services with `runit`.  `socklog` sets itself up. 

You will need to `rc-update add runitd` && `rc-update add socklog` & start the services or reboot the container.

**Comment by candlerb** [comment]
But that's not docker, is it?

**Comment by itoffshore** [comment]
@candlerb - I have [scripts to giver docker fixed ip's](http://it-offshore.co.uk/linux/debian/36-giving-docker-containers-static-ip-addresses) but only use it for a build environment.

**Comment by candlerb** [comment]
Sure, that's pipework. It's mentioned lots in the comments above.

**Comment by netroby** [comment]
Shall we persist the ip for the exists container? for real world example:

Front Web load balance == Link => web server 1 == Link => DB


Once we made some change in web server, we need restart web server , but we don't want to stop front load balance, 
If we can just only restart the web server, we can prevent lots of downtime, and we can make sure our service more smothly

**Comment by cpuguy83** [comment]
@netroby why would you need to stop the lb in this case today?

**Comment by netroby** [comment]
It's running on docker container, because it was link to the web server . if web server restart , the ip changed, the link still valid?

**Comment by cpuguy83** [comment]
Yes, link is still valid.

**Comment by netroby** [comment]
Ok, i got it. but the load balance need to reload , because the ip changes. so i know how to do it.

**Comment by netroby** [comment]
Thanks for your help

**Comment by thaJeztah** [comment]
this may be interesting as well for that https://github.com/ehazlett/interlock

**Comment by candlerb** [comment]
@netroby:

> Front Web load balance == Link => web server 1 == Link => DB
> 
> Once we made some change in web server, we need restart web server , but we don't want to stop front load balance, 

But don't forget the common case where you decide to **upgrade** the web server. With the docker architecture this normally means you have to discard the old container and replace it with a new one.

In that case, there is no benefit to the old container remembering its IP address, since when you start the new container, you will have to instruct it with the correct IP address to use anyway.

Also if you're using "docker link" then you're using private internal addresses behind docker's own NAT, in that case you probably want docker to manage this for you, rather than choosing IP addresses manually for every container you start.

I'm not sure what's the best way to support the use case you suggest (i.e. stopping/starting/recreating a middle tier container for a three-tier architecture), but I don't think containers remembering their assigned IP is the right solution.

**Comment by JonasT** [comment]
> In that case, there is no benefit to the old container remembering its IP address, since when you start the new container, you will have to instruct it with the correct IP address to use anyway.

Why that? I discard and re-create my containers all the time for upgrades (e.g. to have a simple apt-get upgrade run again with newer results), and I almost *never* touch the config let alone manually specify some IP again. Why would recreating a container mean I would want to waste time to manually set some things up? Docker is meant for automation, it should always be designed in a way to not force me to manually do *anything* if there's no good reason for it.

> I'm not sure what's the best way to support the use case you suggest (i.e. stopping/starting/recreating a middle tier container for a three-tier architecture), but I don't think containers remembering their assigned IP is the right solution.

While docker link is a fair solution for a lot of things, it sucks that destroying a linked container and bringing it up again will often result in the other previously connected one to be eternally confused. I don't understand how remembering an assigned IP isn't *the* perfect solution for that. How would it not easily solve that scenario?

**Comment by candlerb** [comment]
> I discard and re-create my containers all the time for upgrades (e.g. to have a simple apt-get upgrade run again with newer results), and I almost never touch the config

I mean settings applied when the container is instantiated, not what's in the Dockerfile.  That is, I thought we were discussing the following proposed feature:

~~~
docker run --ip 1.2.3.4 image1 container1
docker stop container1
docker start container1    # remembers IP address is 1.2.3.4
~~~

Use case: you have four webserver processes behind a load balancer process. You instantiate four copies of the same image, differing only in their IP address. So the above allows you to stop and restart container1 and continue running on the same IP address, which is helpful for the load balancer.

What I'm saying is I don't think that feature will help if you want to destroy container1 and recreate it using image2, because you will have to specify `--ip 1.2.3.4` to the new `docker run` as well. As far as I know, there's no command which says "create container Y but copy all the run-time settings from container X".

If you have to remember externally that the IP address of container1 is 1.2.3.4, you might as well provide `--ip 1.2.3.4` to `docker start` as well, and at that point there is no need to persist the IP address with the container at all.

Now, it might be usefu to persist the IP l if we could do this:

~~~
docker restart --new-image image2 container1
~~~

That of course is a completely different feature request. But even then there are some open questions: does the new container get the same UUID as before, or a new UUID? Does it carry forward all the unionfs layers apart from the base image?

My apologies if I've missed the point here.

**Comment by JonasT** [comment]
@candlerb 
Ah right, I misunderstood before. Yea, I agree with what you're saying.

**Comment by robbrockbank** [comment]
Hi @mavenugo,

I'm looking for a way to specify an IP address when creating a container using `docker run`.  You mentioned that writing an IPAM driver is one way of achieving this.  I have written an IPAM driver, but I'm still not sure how to pass information through to the IPAM driver to know that a particular container needs to be assigned a particular IP address.  What would be the recommended approach when writing your own driver to achieve this.

Is it possible to specify arbitrary options on the `docker run` command that get passed into the IPAM and network drivers, or would you expect some pre-configuration that allows the IPAM driver to lookup a particular container and then select the IP address that it should be assigned, and in this case how would the container endpoint be identified before the container is created?

**Comment by mavenugo** [comment]
@robbrockbank we had a few discussions on this topic recently & @aboch is looking at the way to provide this support in conjunction with IPAM (we had a recent fix in IPAM integration to support that).
@aboch can you please add more details to this ?

**Comment by aboch** [comment]
@robbrockbank 
In order for user to specify a preferred address for the container, plan is to add an explicit `--ip` (`--ip6`) option to `docker run` to be used in conjunction with `--subnet`.
Your ipam driver will ultimately receive the user chosen address as parameter of the `RequestAddress()` call.
This is WIP and it is planned for 1.10.

I am also planning some work to allow DHCP based ipam plugins, where libnetwork would set the endpoint MAC address in the options parameter in the call to `RequestAddress` if the ipam driver expressed this need during registration.

**Comment by kevinxucs** [comment]
@aboch is there a timeframe when 1.10 will be released?

**Comment by thaJeztah** [comment]
@kevinxucs 1.10 is expected to be released somewhere in February, due to the Holidays and New Year

**Comment by Frogging101** [comment]
@aboch The built-in IPAM driver will support this, right? No need to write our own IPAM drivers to specify the IP address when running a container?

**Comment by aboch** [comment]
@Frogging101 
Yes, the built-in IPAM driver will honor the preferred IP option specified by the user.
